crypto/ascon/ascon_aead128.c

```diff
@@ -0,0 +1,436 @@
+        flags = ctx->flags;
+        if (flags & ASCONFLG_AAD) {
+            ctx->flags = 0;
+            ascon_aead128_update(ctx, NULL, &pad, 1);
+            ASCONP8(ctx->state[0], ctx->state[1], ctx->state[2], ctx->state[3], ctx->state[4]);
+            flags ^= ASCONFLG_AAD;
+            ctx->offset = 0;
+        }
+        ctx->state[4] ^= ASCONFLG_DOMAINSEP;
+        flags ^= ASCONFLG_DOMAINSEP;
+        ctx->flags = flags;
+    }
+
+    s0 = ctx->state[0];
+    s1 = ctx->state[1];
+    s2 = ctx->state[2];
+    s3 = ctx->state[3];
+    s4 = ctx->state[4];
+
+    while (len--) {
+        unsigned char ob, ib = *in++;
+
+        if (ctx->offset >= 16) {
+            ASCONP8(s0, s1, s2, s3, s4);
+            ctx->offset = 0;
+        }
+
+        if (ctx->flags & ASCONFLG_DEC) {
+            if (ctx->offset >= 8) {
+                ob = (unsigned char)(s1 >> 8 * (ctx->offset & 0x7)) ^ ib;
+                s1 ^= (uint64_t)(ob) << 8 * (ctx->offset & 0x7);
+            } else {
+                ob = (unsigned char)(s0 >> 8 * ctx->offset) ^ ib;
+                s0 ^= (uint64_t)(ob) << 8 * ctx->offset;
+            }
+        } else {
+            if (ctx->offset >= 8) {
+                s1 ^= (uint64_t)(ib) << 8 * (ctx->offset & 0x7);
+                ob = (unsigned char)(s1 >> 8 * (ctx->offset & 0x7));
+            } else {
+                s0 ^= (uint64_t)(ib) << 8 * ctx->offset;
+                ob = (unsigned char)(s0 >> 8 * ctx->offset);
+            }
+        }
+
+        if (out != NULL)
+            *out++ = ob;
+
+        ctx->offset++;
+    }
+
+    ctx->state[0] = s0;
+    ctx->state[1] = s1;
+    ctx->state[2] = s2;
+    ctx->state[3] = s3;
+    ctx->state[4] = s4;
+}
+
+void ascon_aead128_encrypt_update(ascon_aead128_ctx *ctx, unsigned char *ct,
+                                  const unsigned char *pt, size_t len)
+{
+    ascon_aead128_update(ctx, ct, pt, len);
+}
+
+void ascon_aead128_decrypt_update(ascon_aead128_ctx *ctx, unsigned char *pt,
+                                  const unsigned char *ct, size_t len)
+{
+    ctx->flags |= ASCONFLG_DEC;
+    ascon_aead128_update(ctx, pt, ct, len);
+}
+
+void ascon_aead128_init(ascon_aead128_ctx *ctx, const unsigned char *k,
+                        const unsigned char *n)
+{
+    uint64_t s0, s1, s2, s3, s4, k0, k1;
+
+    GETU64(s1, k);
+    GETU64(s2, k + 8);
+    GETU64(s3, n);
+    GETU64(s4, n + 8);
+    ctx->key[0] = k0 = s1;
+    ctx->key[1] = k1 = s2;
+    s0 = 0x00001000808C0001ULL;
+    ASCONP12(s0, s1, s2, s3, s4);
+    s3 ^= k0;
+    s4 ^= k1;
+    ctx->state[0] = s0;
+    ctx->state[1] = s1;
+    ctx->state[2] = s2;
+    ctx->state[3] = s3;
+    ctx->state[4] = s4;
+    ctx->offset = 0;
+    ctx->flags = ASCONFLG_DOMAINSEP;
+}
+
+void ascon_aead128_aad_update(ascon_aead128_ctx *ctx, const unsigned char *in,
+                              size_t len)
+{
+    uint64_t flags;
+
+    flags = ctx->flags;
+    ctx->flags = 0;
+    ascon_aead128_update(ctx, NULL, in, len);
```

Comment:
Pass a flag to this function rather than doing this save/restore behaviour.

Reply:
upstream, agree to disagree. compiler should inline so it goes away anyway